% Generated by roxygen2 (4.1.1): do not edit by hand
% Please edit documentation in R/gbt.default.R
\name{gbt.default}
\alias{gbt.default}
\title{Create new gbt object}
\usage{
\method{gbt}{default}(covstats, mark = NA, marksource = NA, ssu = NA,
  trna = NA, dec = getOption("OutDec"))
}
\arguments{
\item{covstats}{File(s) with coverage statistics of a metagenomic assembly;
output of pileup.sh in BBTools package (required). More
than one file can be specified with c().}

\item{mark}{Table of scaffolds with marker genes and taxonomic information
for each marker. E.g. AMPHORA2 or Phyla-AMPHORA marker sets
and output parsed by parse_phylotype_result.pl. (optional)}

\item{ssu}{Table of scaffolds with SSU rRNA genes, and taxonomic info
for each SSU. E.g. use get_ssu_for_genome_bin_tools.pl.
(optional)}

\item{dec}{Decimal separator used in covstats tables. Default is the
           current locale default found with getOption("OutDec")}

\item{tnra}{Table of tRNA genes found in assembly. Can use the output from
tRNAscan-SE directly. (optional)}
}
\value{
Object of class gbt
}
\description{
Creates new gbt object from coverage data, taxonomic markers, and other
data
}
\details{
See documentation online https://github.com/kbseah/genome-bin-tools for
fuller instructions on generating the input files required.
}
\seealso{
\code{\link{gbtbin}}, \code{\link{choosebin}}
}

